lib/file_browser/widgets/file_browser.dart

```diff
@@ -0,0 +1,167 @@
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+import 'package:flutter/cupertino.dart';
+import 'package:provider/provider.dart';
+import 'package:schulcloud/app/app.dart';
+import 'package:schulcloud/courses/courses.dart';
+
+import '../bloc.dart';
+import '../data.dart';
+import 'app_bar.dart';
+import 'page_route.dart';
+
+const _loadingContent = const Align(
+  alignment: Alignment.topCenter,
+  child: Padding(
+    padding: EdgeInsets.only(top: 32),
+    child: CircularProgressIndicator(),
+  ),
+);
+
+class FileBrowser extends StatelessWidget {
+  final Entity owner;
+  final File parent;
+
+  /// Whether this widget is embedded into another screen. If [true], doesn't
+  /// show an app bar.
+  final bool showAppBar;
+
+  FileBrowser({
+    @required this.owner,
+    this.parent,
+    this.showAppBar = true,
+  })  : assert(owner != null),
+        assert(owner is Course || owner is User),
+        assert(parent == null || parent.isDirectory),
+        assert(showAppBar != null);
+
+  Course get ownerAsCourse => owner is Course ? owner as Course : null;
+
+  void _openDirectory(BuildContext context, File file) {
+    assert(file.isDirectory);
+
+    Navigator.of(context).push(FileBrowserPageRoute(
+      builder: (context) => FileBrowser(owner: owner, parent: file),
+    ));
+  }
+
+  Future<void> _downloadFile(BuildContext context, File file) async {
+    assert(file.isNotDirectory);
+
+    Scaffold.of(context).showSnackBar(SnackBar(
+      content: Text('Downloading file ${file.name}'),
+    ));
+    try {
+      await Provider.of<Bloc>(context).downloadFile(file);
+    } on PermissionNotGranted catch (_) {
+      Scaffold.of(context).showSnackBar(SnackBar(
```

Comment:
This `SnackBar` is not shown on my device without storage permissions. Maybe move the other one into the try block so that only one will be shown?

Reply:
Done. The problem was that I checked for the `permission.value` to be `null` instead of `0`. Now, I also changed it so the "Downloading [file]" Snackbar only gets shown when the file is actually downloaded. The other Snackbar ("To download files, we need to access your storage.") now offers a button for granting that persmission.